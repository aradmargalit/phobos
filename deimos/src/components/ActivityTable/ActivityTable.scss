.table-edit-buttons {
  display: flex;
  justify-content: space-between;
}

.search-bar {
  width: 50%;
}
