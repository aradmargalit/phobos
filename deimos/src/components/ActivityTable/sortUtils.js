export const numberSorter = (a, b) => a.logical_index - b.logical_index;

export const nameSorter = (a, b) => a.name.localeCompare(b.name);

export const dateSorter = (a, b) => a.epoch - b.epoch;

export const activityTypeSorter = (a, b) =>
  a.activity_type.name.localeCompare(b.activity_type.name);

export const durationSorter = (a, b) => a.duration - b.duration;

export const distanceSorter = (a, b) => a.distance - b.distance;
