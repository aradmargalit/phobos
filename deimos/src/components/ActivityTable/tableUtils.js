import { snakeCase as _snakeCase, uniq as _uniq } from 'lodash';

import { formatDate } from '../../utils/dataFormatUtils';

export const toCol = (name, render) => {
  const snakeName = _snakeCase(name);
  const col = { title: name, dataIndex: snakeName, ellipsis: true };
  if (render) {
    col.render = render;
  }
  return col;
};

export const filterActivities = (term, activities) => {
  if (!term || term.length === 0) return activities;

  const filteredByName = activities.filter((ta) =>
    ta.name.toLowerCase().includes(term.toLowerCase())
  );

  const filteredByType = activities.filter((ta) =>
    ta.activity_type.name.toLowerCase().includes(term.toLowerCase())
  );

  const filteredByDate = activities.filter((ta) =>
    formatDate(ta.activity_date).toLowerCase().includes(term.toLowerCase())
  );

  return _uniq([...filteredByName, ...filteredByType, ...filteredByDate]);
};
