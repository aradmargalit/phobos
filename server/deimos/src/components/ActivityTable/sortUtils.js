import moment from 'moment';

export const numberSorter = (a, b) => {
  return a.logical_index - b.logical_index;
};

export const nameSorter = (a, b) => {
  return a.name.localeCompare(b.name);
};

export const dateSorter = (a, b) => {
  return moment(a.activity_date).diff(moment(b.activity_date));
};

export const activityTypeSorter = (a, b) => {
  return a.activity_type.name.localeCompare(b.activity_type.name);
};

export const durationSorter = (a, b) => {
  return a.duration - b.duration;
};

export const distanceSorter = (a, b) => {
  return a.distance - b.distance;
};
