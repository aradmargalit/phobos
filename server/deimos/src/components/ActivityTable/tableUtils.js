import { snakeCase as _snakeCase, uniq as _uniq } from 'lodash';

import { formatDate } from '../../utils/dataFormatUtils';

export const toCol = (name, render) => {
  const snakeName = _snakeCase(name);
  const col = { title: name, dataIndex: snakeName, ellipsis: true };
  if (render) {
    col.render = render;
  }
  return col;
};

export const filterActivities = (term, activities) => {
  if (!term || term.length === 0) return activities;

  const filteredByName = activities.filter(ta =>
    ta.name.toLowerCase().includes(term)
  );

  const filteredByType = activities.filter(ta =>
    ta.activity_type.name.toLowerCase().includes(term)
  );

  const filteredByDate = activities.filter(ta =>
    formatDate(ta.activity_date)
      .toLowerCase()
      .includes(term)
  );

  return _uniq([...filteredByName, ...filteredByType, ...filteredByDate]);
};
