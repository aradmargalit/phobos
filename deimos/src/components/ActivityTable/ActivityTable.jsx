import './ActivityTable.scss';

import { DeleteOutlined, EditOutlined } from '@ant-design/icons';
import { Button, Empty, Input, Modal, Popconfirm, Spin, Table } from 'antd';
import { debounce as _debounce } from 'lodash';
import moment from 'moment';
import React, { useState } from 'react';

import { deleteActivity } from '../../apis/phobos-api';
import { formatDate, minutesToHMS } from '../../utils/dataFormatUtils';
import AddActivityForm from '../AddActivityForm';
import { filterActivities, makeDurationBreakdown, toCol } from './tableUtils';

const { Search } = Input;

export default function ActivityTable({ loading, activities, refetch }) {
  const [editModalVisible, setEditModalVisible] = useState(false);
  const [editingActivity, setEditingActivity] = useState(null);
  const [searchTerm, setSearchTerm] = useState(null);

  if (loading) return <Spin />;
  if (!activities) return <Empty description="No activities...yet!" />;

  const bouncedSetSearchTerm = _debounce(setSearchTerm, 500);

  const onChangeHandler = e => {
    if (!e.target.value || !e.target.value.length) {
      setSearchTerm(null);
      return;
    }
    bouncedSetSearchTerm(e.target.value);
  };

  const renderEditButtons = activity => (
    <Button
      onClick={() => {
        const toEdit = { ...activity };
        toEdit.activity_date = moment(activity.activity_date);
        // This is dumb, but we need to set the full duration object until I get smarter
        toEdit.duration = makeDurationBreakdown(activity.duration);
        setEditingActivity(toEdit);
        setEditModalVisible(true);
      }}
    >
      Edit
    </Button>
  );

  const confirmDelete = ({ id }) => (
    <Popconfirm
      title="Are you sure?"
      okText="Delete"
      icon={<DeleteOutlined style={{ color: 'red' }} />}
      onConfirm={async () => {
        await deleteActivity(id);
        refetch();
      }}
    >
      <Button ghost type="danger">
        Delete
      </Button>
    </Popconfirm>
  );

  const closeModal = () => setEditModalVisible(false);
  const columns = [
    {
      title: 'No.',
      dataIndex: 'idx',
      render: (text, record, idx) => activities.length - idx,
    },
    { ...toCol('Name'), width: 250 },
    // We want to format this one as the time it was entered, since it's time is 00:00:00
    // and we don't want to cross date boundaries by converting timezones
    toCol('Activity Date', formatDate),
    {
      title: 'Activity Type',
      dataIndex: ['activity_type', 'name'],
    },
    toCol('Duration', minutesToHMS),
    toCol('Distance', (distance, record) =>
      distance > 0 ? `${distance} ${record.unit}` : '-'
    ),
    {
      title: <EditOutlined />,
      key: 'edit',
      align: 'center',
      render: renderEditButtons,
    },
    {
      title: <DeleteOutlined />,
      key: 'delete',
      align: 'center',
      render: confirmDelete,
    },
  ];

  return (
    <div>
      <Search
        className="search-bar"
        allowClear
        placeholder="Search by name, type, or date..."
        onSearch={onChangeHandler}
        onChange={onChangeHandler}
        width="50%"
      />
      <Table
        rowKey="id"
        dataSource={filterActivities(searchTerm, activities)}
        columns={columns}
      />
      <Modal
        title="Edit Activity"
        visible={editModalVisible}
        onOk={closeModal}
        onCancel={closeModal}
        width={750}
        footer={null}
        destroyOnClose
      >
        <AddActivityForm
          refetch={refetch}
          initialActivity={editingActivity}
          modalClose={closeModal}
        />
      </Modal>
    </div>
  );
}
