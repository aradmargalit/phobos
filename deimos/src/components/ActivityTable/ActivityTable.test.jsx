import { shallow } from 'enzyme';
import React from 'react';

import ActivityTable from './ActivityTable';

const generateComponent = props => {
  const defaultProps = {};
  const mergedProps = { ...defaultProps, ...props };

  return shallow(<ActivityTable {...mergedProps} />);
};

describe('<ActivityTable />', () => {
  it('renders correctly', () => {
    expect(generateComponent());
  });

  describe('while loading', () => {
    const component = generateComponent({ loading: true });
    it('has a spinner and no data', () => {
      expect(component.find('Spin')).toHaveLength(1);
      expect(component.find('Empty')).toHaveLength(0);
    });
  });

  describe('with no activities...yet', () => {
    it('renders an Empty component', () => {
      expect(generateComponent().find('Empty')).toHaveLength(1);
    });
  });

  describe('with activities', () => {
    it('renders a table with no pagination', () => {
      const component = generateComponent({
        activities: [
          {
            strava_id: 1,
            logical_index: 1,
            name: 'Fun Run',
            activity_date: new Date('2020-01-01'),
            activity_type: { name: 'Run' },
            duration: 12,
            distance: 40,
          },
        ],
      });
      const table = component.find('Table');
      expect(table).toHaveLength(1);
      expect(table.prop('dataSource')).toHaveLength(1);
      expect(table.prop('columns')).toHaveLength(10);
    });
  });
});
