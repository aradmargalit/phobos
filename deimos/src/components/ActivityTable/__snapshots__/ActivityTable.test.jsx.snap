// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActivityTable /> renders correctly 1`] = `
<Empty
  description="No activities...yet!"
/>
`;
